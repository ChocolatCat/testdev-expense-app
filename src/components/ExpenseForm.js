import React from 'react';
import DatePicker from 'react-datepicker';

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            createdAt: props.expense ? props.expense.createdAt : Date.now(),
            calendarFocused: false,
            error: ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>({
            description
        }));
    }
    onTextareaChange = (e) => {
        const note = e.target.value;
        this.setState(()=>({
            note
        }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        //Regex for a number with 2 decimal numbers
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({
                amount
            }));
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({
                createdAt: Date.parse(createdAt)
            }));
        }
    }
    onSubmit = (e) => {
        //prevent refresh
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            //Set error state = to "Please provide description and amount"
            this.setState( () => ({
                error: "Please provide a description and amount!"
            }));
        }
        else{
            //clear the error
            this.setState( () => ({
                error: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt
            });
        }
    }
    render(){
        return (
            <div>
                {this.state.error && <h1>{this.state.error}</h1>}
                <form name="expenseForm" onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    >
                    </input>
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    >
                    </input>
                    <DatePicker 
                        selected={this.state.createdAt}
                        onChange={this.onDateChange}
                        dateFormat="dd/MM/yyyy"
                    />
                    <textarea
                        placeholder="Add a note to this expense! (optional)"
                        value={this.state.note}
                        onChange={this.onTextareaChange}
                    >
                    </textarea>
                    <button name="submit">
                        Add Expense
                    </button>
                </form>
            </div>
        );
    }
}