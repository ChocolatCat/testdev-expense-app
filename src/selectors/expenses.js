//import {createSelector} from '@reduxjs/toolkit';
import {isAfter, isBefore, isSameDay} from 'date-fns';

const selectExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = expense.createdAt;
        const startDateMatch = startDate ?  isAfter(createdAtMoment, startDate) || isSameDay(startDate, createdAtMoment) : true;
        const endDateMatch = endDate ? isBefore(createdAtMoment, endDate) || isSameDay(endDate, createdAtMoment) : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt > b.createdAt ? 1 : -1
        }
        if(sortBy === 'amount'){
            return a.amount > b.amount ? 1 : -1;
        }
    });  
};

const selectExpenseById = (expenses, id) => {
    return expenses.find( (expense) => {
        return expense.id === id;
    });
};

const selectExpensesTotal = (expenses) => {
    return expenses
        .map( (expense) => expense.amount )
        .reduce( (sum, value) => sum+value, 0);
}

export {selectExpenses, selectExpenseById, selectExpensesTotal};