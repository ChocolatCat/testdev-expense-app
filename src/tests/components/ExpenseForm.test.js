/** @jest-environment jsdom */

import React from 'react';
import {render, fireEvent} from '@testing-library/react';
import ExpenseForm from '../../components/ExpenseForm';
import { expenses } from '../fixtures/expenses';

//require.requireActual to ask for the real library

jest.useFakeTimers().setSystemTime(new Date(2022, 7, 10, 8));


it('Should render ExpenseForm', () => {
    const {asFragment} = render(
        <ExpenseForm />
    );
    expect(asFragment()).toMatchSnapshot();
});

it('Should render ExpenseForm with data', () => {
    const {asFragment} = render(
        <ExpenseForm expense={expenses[1]}/>
    );
    expect(asFragment()).toMatchSnapshot();
});

it('Should render error for invalid form submission', () => {
    const {getByRole, getByText} = render(
        <ExpenseForm />
    );
    fireEvent.submit(getByRole('form'));
    expect(getByText("Please provide a description and amount!")).toBeInTheDocument();
});

it('Should set description on input change', () => {
    const {getByPlaceholderText} = render(
        <ExpenseForm />
    );
    const description = getByPlaceholderText("Description");
    fireEvent.change(description, {target: {value:"Todokete"}});
    expect(description.value).toBe("Todokete");
});

it('Should set note on textarea change', () => {
    const {getByPlaceholderText} = render(
        <ExpenseForm />
    );
    const note = getByPlaceholderText("Add a note to this expense! (optional)");
    fireEvent.change(note, {target: {value:"Todokete"}});
    expect(note.value).toBe("Todokete");
});

it('Should set amount to valid input', () => {
    const {getByPlaceholderText} = render(
        <ExpenseForm />
    );
    const amount = getByPlaceholderText("Amount");
    fireEvent.change(amount, {target: {value:"12345"}});
    expect(amount.value).toBe("12345");
});

it('Should not set amount to invalid input', () => {
    const {getByPlaceholderText} = render(
        <ExpenseForm />
    );
    const amount = getByPlaceholderText("Amount");
    fireEvent.change(amount, {target: {value:"5a2"}});
    expect(amount.value).toBe("");
});

it('should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn(); //We create a spy
    const {getByRole} = render(
        <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy}/>
    );
    fireEvent.submit(getByRole('form'));
    expect(onSubmitSpy).toHaveBeenCalledWith({
        description: expenses[0].description,
        amount: expenses[0].amount,
        note: expenses[0].note,
        createdAt: expenses[0].createdAt
    });
});

it('should call onDateChange', ()=> {
    const onSubmitSpy = jest.fn(); //We create a spy
    const {getAllByRole} = render(
        <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy}/>
    );
    const datePicker = getAllByRole('textbox')[2];
    fireEvent.mouseDown(datePicker);
    //american date
    fireEvent.change(datePicker, { target: { value: "12-15-2022" } });
    //formated date
    expect(datePicker.value).toBe('15/12/2022');
});