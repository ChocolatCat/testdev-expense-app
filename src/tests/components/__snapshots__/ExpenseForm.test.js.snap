// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseForm 1`] = `
<DocumentFragment>
  <div>
    <form
      name="expenseForm"
    >
      <input
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        placeholder="Amount"
        type="text"
        value=""
      />
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            type="text"
            value="10/08/2022"
          />
        </div>
      </div>
      <textarea
        placeholder="Add a note to this expense! (optional)"
      />
      <button
        name="submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Should render ExpenseForm with data 1`] = `
<DocumentFragment>
  <div>
    <form
      name="expenseForm"
    >
      <input
        placeholder="Description"
        type="text"
        value="Rent"
      />
      <input
        placeholder="Amount"
        type="text"
        value="40000"
      />
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            type="text"
            value="06/08/2022"
          />
        </div>
      </div>
      <textarea
        placeholder="Add a note to this expense! (optional)"
      >
        For august
      </textarea>
      <button
        name="submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;
